index.ts

```diff
@@ -211,22 +211,49 @@ const getReversedInfoIterator = function *(info: ObjectInfo): IterableIterator<O
 export const getPath = (info: ObjectInfo): ReadonlyArray<string|number> =>
     _.reverse(_.filterMap(getReversedInfoIterator(info), i => i.isChild ? i.property : undefined))
 
-export const cloneDeep = <T extends Data>(source: T): T => {
-    const data: Data = source
-    if (data === null ||
-        typeof data === "boolean" ||
-        typeof data === "number" ||
-        typeof data === "string"
-    ) {
-        return source
+/**
+ * Returns a deep clone of `source` and set a source-map for each member.
+ *
+ * @param source an original object
+ * @param getOptional the function should return an object info of a provided member.
+ *                    If the function is not provided the algorithm extract information from
+ *                    the provided member.
+ */
+export const cloneDeep = <T extends Data>(
+    source: T,
+    getOptional?: (member: Data|undefined) => InfoFunc|undefined
```

Comment:
rename to something more explicit on what it does ( i,e, that it overrides the existing info